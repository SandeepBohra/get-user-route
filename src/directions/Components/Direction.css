.Direction {
    margin-left: 10px;
}

.location-details-form {
    width: 25%;
    height: 100%;
}

#map {
    margin: 10px;
    top: 0%;
    left: 25%;
    position: absolute;
    width: 70%;
    height: 90%;
}

.loader-container {
    position: absolute;
    left: 0;
    top: 0;
    background-color: #80808038;
    width: 100%;
    height: 100%;
    z-index: 999;
  }
  .loader-container > .loader {
    position: absolute;
    margin: auto;
    left: 0;
    right: 0;
    bottom: 0;
    top: 0;
    border: 6px solid #f3f3f3;
    border-top: 6px solid #000;
    border-radius: 50%;
    width: 2em;
    height: 2em;
    animation: spin 2s linear infinite;
  }
  
  @keyframes spin {
    0% {
      transform: rotate(0deg);
    }
    100% {
      transform: rotate(360deg);
    }
  }
  

