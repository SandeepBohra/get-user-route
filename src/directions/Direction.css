.Direction {
    margin-left: 10px;
}

.location-details-form {
    width: 25%;
    height: 100%;
}

.error-container {
  margin-top: 30px;
  color: red;
  font-size: 16px;
  font-weight: 500;
}

.map-container > .map {
        margin: 10px;
        top: 0%;
        left: 25%;
        position: absolute;
        width: 70%;
        height: 90%;
}