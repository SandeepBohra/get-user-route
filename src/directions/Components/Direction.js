import React, { Component } from 'react';
import LocationDetailsForm from './LocationDetailsForm/LocationDetailsForm';
import { googleMaps } from '../Services/googleMap'
import { getTokenAndRoute } from '../Services/getUserRoute'
import './Direction.css';


class Direction extends Component {
    mapContainer
    constructor() {
        super();
        this.localMapVariable = {};
        this.renderMap = {};
        this.state = {
            isLoading: false,
            routeReceived: null,
            showRouteDistAndTime: false,
            routeDistance: '',
            routeTime: '',
        }
    }
    componentDidMount() {
        this.initializeGoogleMaps();
    }
    
    initializeGoogleMaps = async () => {
        const mapSettings = {
            zoom: 12,
            center: { lat: 22.372081, lng: 114.107877 }
        }
        const { googleMaps } = this.props
        this.localMapVariable = await googleMaps();
        this.renderMap = new this.localMapVariable.Map(this.mapContainer, mapSettings);
    }

    drawRouteOnMap = (path) => {
        const directionService = new this.localMapVariable.DirectionsService();
        const directionRenderer = new this.localMapVariable.DirectionsRenderer();
        const origin = path[0];
        const dest = path[path.length -1];
        directionRenderer.setMap(this.renderMap);
        directionService.route({
            origin: new this.localMapVariable.LatLng(origin[0], origin[1]),
            destination: new this.localMapVariable.LatLng(dest[0], dest[1]),
            travelMode: "DRIVING"
            },  (response, status) => {
            if (status === 'OK') {
                directionRenderer.setDirections(response);
            }
        });
    }

    showLoader = (set) => {
        this.setState({
            isLoading: set
        })
    }

    sendLocationAndGetRoute = async (orig, dest) => {
        this.showLoader(true)
        const response = await getTokenAndRoute(orig, dest);
        this.showLoader(false)
        if(response.data && response.data.path) {
            this.setState({
                showRouteDistAndTime: true,
                routeDistance: response.data.total_distance,
                routeTime: response.data.total_time
            })
            const { path } = response.data
            this.drawRouteOnMap(path);
        } else if(response.data && response.data.error) {
            alert(response.data.error);
        }
    }

    renderLoader = () => {
        return (
            <div className="loader-container">
                <div className="loader" />
            </div>
        )
    }

    render() {
        return (
            <div className="Direction">
                <div className="location-details-form">
                {this.state.isLoading 
                ? this.renderLoader()
                : null}
                <LocationDetailsForm
                    sendLocationAndGetRoute={this.sendLocationAndGetRoute}
                    showRouteDistAndTime={this.state.showRouteDistAndTime}
                    routeDistance={this.state.routeDistance}
                    routeTime={this.state.routeTime}
                />
                </div>
                <div className="map-container">
                    <div id='map' ref={elem => (this.mapContainer = elem)}></div>
                </div>
            </div>
        )
    }
}

export default Direction;

Direction.defaultProps = {
    googleMaps,
}