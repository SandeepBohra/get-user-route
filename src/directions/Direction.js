import React, { Component } from 'react';
import LocationDetailsForm from './Components/LocationDetailsForm/LocationDetailsForm';
import { googleMaps } from './Services/googleMap';
import { getUserRouteAndToken } from './Services/mockAPIService/index';
import { MOCK_API_ERROR } from './Constants/errorTypes';
import { Loader, MapComponent } from '../Shared/Components/index'
import './Direction.css';


class Direction extends Component {
    mapContainer
    constructor() {
        super();
        this.googleMaps = {};
        this.renderMap = {};
        this.state = {
            isLoading: false,
            showRouteDistAndTime: false,
            routeDistance: '',
            routeTime: '',
            errorMsg: '',
        }
    }
    componentDidMount() {
        this.initializeGoogleMaps();
    }
    
    initializeGoogleMaps = async () => {
        const { googleMaps } = this.props
        this.googleMaps = await googleMaps();
        this.renderMap = new this.googleMaps.Map(this.mapContainer, this.props.mapSettings);
    }

    drawRouteOnMap = (path) => {
        const directionService = new this.googleMaps.DirectionsService();
        const directionRenderer = new this.googleMaps.DirectionsRenderer();
        const origin = path[0];
        const dest = path[path.length -1];
        directionRenderer.setMap(this.renderMap);
        directionService.route({
            origin: new this.googleMaps.LatLng(origin[0], origin[1]),
            destination: new this.googleMaps.LatLng(dest[0], dest[1]),
            travelMode: "DRIVING"
            },  (response, status) => {
            if (status === 'OK') {
                directionRenderer.setDirections(response);
            }
        });
    }

    showLoader = (set) => {
        this.setState({
            isLoading: set
        })
    }

    clearPerviousRoueteDetails = () => {
        this.setState({
            errorMsg: '',
            routeDistance: '',
            routeTime: '',
            showRouteDistAndTime: false,
        })
    }
    sendLocationAndGetRoute = async (orig, dest) => {
        this.showLoader(true)
        this.clearPerviousRoueteDetails()
        const response = await getUserRouteAndToken(orig, dest).catch(e => {
            this.setState({
                errorMsg: MOCK_API_ERROR,
            })
        });;
        this.showLoader(false)
        if(response && response.data.path) {
            this.setState({
                showRouteDistAndTime: true,
                routeDistance: response.data.total_distance,
                routeTime: response.data.total_time
            })
            const { path } = response.data
            this.drawRouteOnMap(path);
        } else if(response && response.data.error) {
            this.setState({
                errorMsg: response.data.error
            })
        }
    }

    resetExistingRouteInformation = () => {
        this.clearPerviousRoueteDetails();
        this.initializeGoogleMaps();
    }

    render() {
        return (
            <div className="Direction">
                <div className="location-details-form">
                {this.state.isLoading 
                ? <Loader />
                : null}
                <LocationDetailsForm
                    sendLocationAndGetRoute={this.sendLocationAndGetRoute}
                    showRouteDistAndTime={this.state.showRouteDistAndTime}
                    routeDistance={this.state.routeDistance}
                    routeTime={this.state.routeTime}
                    resetExistingRouteInformation={this.resetExistingRouteInformation}
                />
                </div>
                <div className="error-container">
                    {this.state.errorMsg}
                </div>
                <MapComponent 
                    fieldRef={elem => (this.mapContainer = elem)}
                />
            </div>
        )
    }
}

export default Direction;

Direction.defaultProps = {
    googleMaps,
    mapSettings: {
        zoom: 12,
        center: { lat: 22.372081, lng: 114.107877 }
    }
}