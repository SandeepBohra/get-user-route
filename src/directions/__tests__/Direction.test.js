import React from 'react';
import renderer from 'react-test-renderer';
import Direction from '../Direction';

describe('<Direction /> component', () => {
  
  it('should match the snapshot', () => {
    const output = renderer.create(<Direction />).toJSON();
    expect(output).toMatchSnapshot();
  })
})