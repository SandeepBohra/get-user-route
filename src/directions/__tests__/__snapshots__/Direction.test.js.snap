// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Direction /> component should match the snapshot 1`] = `
<div
  className="direction"
>
  <div
    className="location-details-form"
  >
    <div
      className="location-details-form"
    >
      <div
        className="input-places-autocomplete"
      >
        <p>
          Strarting location
        </p>
        <input
          name="startingLocation"
          onChange={[Function]}
          type="text"
        />
      </div>
      <div
        className="input-places-autocomplete"
      >
        <p>
          Drop-off point
        </p>
        <input
          name="dropOffPoint"
          onChange={[Function]}
          type="text"
        />
      </div>
      <div
        className="form-buttons"
      >
        <button
          disabled={true}
          onClick={[Function]}
          style={
            Object {
              "marginRight": "30%",
            }
          }
        >
          Submit
        </button>
        <button
          onClick={[Function]}
        >
          Reset
        </button>
      </div>
    </div>
  </div>
  <div
    className="error-container"
  >
    
  </div>
  <div
    className="map-container"
  >
    <div
      className="map"
    />
  </div>
</div>
`;
